File: FirstLastPositionInSortedArray.java
//TC: O(log n), two binary searches
//SC: no extra space - O(1)

//approach: first doing one binary search over array to find first position of target
// second binary search to find last position of target
public class FirstLastPositionInSortedArray {
    public int[] searchRange(int[] nums, int target) {
        if(null == nums || nums.length ==0 ||
                target < nums[0] || target > nums[nums.length -1])
            return new int[] {-1,-1};

        int firstPosition = binarySearchFirst(nums, target);
        if(firstPosition == -1) return new int[] {-1,-1};
        int lastPosition = binarySearchLast(nums, target);
        return new int[] {firstPosition, lastPosition};
    }

    int binarySearchFirst(int[] nums, int target){
        int low = 0; int high = nums.length-1;
        while(low<=high){
            int mid = low + (high-low)/2;
            if(nums[mid] == target){
                if(mid == low || nums[mid-1] != target) return mid;
                else high = mid-1;
            } else if(nums[mid] < target){
                low = mid+1;
            } else{
                high = mid-1;
            }
        }
        return -1;
    }

    int binarySearchLast(int[] nums, int target){
        int low = 0; int high = nums.length-1;
        while(low<=high){
            int mid = low + (high-low)/2;
            if(nums[mid] == target){
                if(mid == high || nums[mid+1] != target) return mid;
                else low = mid+1;
            }else if(nums[mid] < target){
                low = mid + 1;
            }else{
                high = mid - 1;
            }
        }
        return -1;
    }
}

File: first_last_position_in_sorted_array.py
# Time Complexity = O(log n)
# Sapce Complexity = O(1)
# Leetcode = Yes
# Approach = We can perform two binary searches, StartPos and EndPos, to find first and last occurance. StartPos finds the first occurrence of the target by checking if the current mid is the start or if the previous element is smaller. Similarly, EndPos finds the last occurrence by checking if the current mid is the end or if the next element is larger. Both functions run in 


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        class FindTargetRange:
            def StartPos(arr, target, low, high):
                while low <= high:
                    mid = low + (high - low) // 2
                    if arr[mid] == target:
                        if mid == low or arr[mid - 1] < arr[mid]:
                            return mid
                        else:
                            high = mid - 1
                    elif arr[mid] > target:
                        high = mid - 1
                    else:
                        low = mid + 1
                return -1

            def EndPos(arr, target, low, high):
                while low <= high:
                    mid = low + (high - low) // 2
                    if arr[mid] == target:
                        if mid == high or arr[mid] < arr[mid + 1]:
                            return mid
                        else:
                            low = mid + 1
                    elif arr[mid] > target:
                        high = mid - 1
                    else:
                        low = mid + 1
                return -1

        firstPos = FindTargetRange.StartPos(nums, target, 0, len(nums) - 1)
        lastPos = FindTargetRange.EndPos(nums, target, 0, len(nums) - 1)
        return [firstPos, lastPos]
